future-proof-iot/RIOT
# sys: add suitreg module

### Contribution description



This PR adds an initial implementation for suitreg module. This module provides a way for thread to registering to suit process flow. There are three type of registrations: STATUS, ERRORS and BLOCK. STATUS and ERRORS will just mean getting a message about the suit workflow progress, when updates are received, how they are progressing and if any errors are encountered but without blocking it. The messages can deliver specif content about the update process: like the size of the firmware to download and the amount of firmware already downloaded.

When subscribing to BLOCK messages will be sent in a msg_send_receive way so the subscribing thread can halt the progress of the update. This is done before initiating intensive cpu or networking operations: verifying the manifest and digest, or initiating the download: as well as before the actual reboot.

Its been implemented completely decoupled from the rest of suit. For now I'm using this to be able to follow the update process in `future-proof-iot/demo`, but I think its useful to have something like this for an application to delay a reboot or delay the update process.

It would be nice for me to get this into suit-dev to be able to use it cleanly in the demo, but it doesn't need to got into master right away. Anyway hopping for thoughts on the concept.

### Testing procedure



Run unittest:  

`make -C tests/unittests/ tests-suitreg all term`

In an application running along suit register:

```
    suit_reg_t entry = SUIT_REG_INIT_PID(SUIT_REG_TYPE_STATUS, thread_getpid());
    suit_reg_register(&entry);
```

Do something about the messages:

```
    while (1) {
        msg_receive(&m);
        switch(m.type) {
            case SUIT_TRIGGER:
                printf("Update trigger received \n")
                break;
            case SUIT_SIGNATURE_START:
                printf("Verifying signature \n")
                break;
            case SUIT_REBOOT:
                printf("Going to reboot \n")
                break;
            case SUIT_DOWNLOAD_FW:
                printf("Firmware size: %lu \n", m.content.value)
                break;
            case SUIT_DOWNLOAD_PROGRESS:
                printf("Downloaded bytes: %lu \n", m.content.value)
                break;
            default:
                break;
        }
    }
```

### Issues/PRs references
